audio/ffmpeg.py

```diff
@@ -0,0 +1,200 @@
+
+        self.ffmpeg_binary = ffmpeg_binary if ffmpeg_binary else "ffmpeg"
+
+        self.recover_duration = recover_duration
+        self.output_format = output_format
+
+        self.rqmt = {"time": 4, "cpu": 4, "mem": 8}
+
+    def tasks(self):
+        yield Task("run", rqmt=self.rqmt)
+        if self.recover_duration:
+            # recovering is not multi-threaded, so force cpu=1
+            recover_rqmt = copy.copy(self.rqmt)
+            recover_rqmt["cpu"] = 1
+            yield Task("run_recover_duration", rqmt=recover_rqmt)
+
+    def run(self):
+        in_corpus = corpus.Corpus()
+        out_corpus = corpus.Corpus()
+
+        in_corpus.load(tk.uncached_path(self.corpus_file))
+        out_corpus.name = in_corpus.name
+        out_corpus.speakers = in_corpus.speakers
+        out_corpus.default_speaker = in_corpus.default_speaker
+        out_corpus.speaker_name = in_corpus.speaker_name
+
+        # store index of last segment
+        for r in in_corpus.recordings:
+            nr = corpus.Recording()
+            nr.name = r.name
+            nr.segments = r.segments
+            nr.speaker_name = r.speaker_name
+            nr.speakers = r.speakers
+            nr.default_speaker = r.default_speaker
+
+            audio_name = r.audio.split("/")[-1]
+
+            if self.output_format is not None:
+                name, ext = os.path.splitext(audio_name)
+                audio_name = name + "." + self.output_format
+
+            nr.audio = os.path.join(tk.uncached_path(self.audio_folder), audio_name)
+            out_corpus.add_recording(nr)
+
+        from multiprocessing import pool
+
+        p = pool.Pool(self.rqmt["cpu"])
+        p.map(self._perform_ffmpeg, in_corpus.recordings)
+
+        if self.recover_duration:
+            out_corpus.dump(tk.uncached_path("temp_corpus.xml.gz"))
+        else:
+            out_corpus.dump(tk.uncached_path(self.out))
+
+    def run_recover_duration(self):
+        """
+        Open all files with "soundfile" and extract the length information
+
+        :return:
+        """
+        import soundfile
+
+        c = corpus.Corpus()
+        c.load("temp_corpus.xml.gz")
+
+        for r in c.all_recordings():
+            assert len(r.segments) == 1, "needs to be a single segment recording"
+            old_duration = r.segments[0].end
+            data, sample_rate = soundfile.read(open(r.audio, "rb"))
+            new_duration = len(data) / sample_rate
+            print(
+                "%s: adjusted from %f to %f seconds"
+                % (r.segments[0].name, old_duration, new_duration)
+            )
+            r.segments[0].end = new_duration
+
+        c.dump(tk.uncached_path(self.out))
+
+    def _perform_ffmpeg(self, recording):
+        """
+        Build and call an FFMPEG command to apply on a recording
+
+        :param corpus.Recording recording:
+        :return:
+        """
+        audio_name = recording.audio.split("/")[-1]
```

Comment:
I know some corpora which have a nested hierarchy of directories for the recordings (e.g. `A/1.mp3`, `B/1.mp3`, etc), so just picking the base name (`1.mp3`) would break this.

Reply:
This is why I already put a warning into the docstring

Reply:
Yes, but why not just fix it?

Reply:
To expand: One solution I commonly do in such cases: Find the common base path of all audio filenames. And then strip away this common base path.